browser-extension/mv2/src/client/js/ruleExecutionHandler.js

```diff
@@ -6,30 +6,35 @@ RQ.RuleExecutionHandler.setup = () => {
   chrome.runtime.onMessage.addListener((message, _, sendResponse) => {
     switch (message.action) {
       case RQ.CLIENT_MESSAGES.UPDATE_APPLIED_RULE_ID:
+        const isFirstExecution = !RQ.RuleExecutionHandler.appliedRuleIds.has(message.ruleId);
+
         RQ.RuleExecutionHandler.appliedRuleIds.add(message.ruleId);
+        sendResponse(isFirstExecution);
         break;
 
       case RQ.CLIENT_MESSAGES.GET_APPLIED_RULE_IDS:
         sendResponse(Array.from(RQ.RuleExecutionHandler.appliedRuleIds));
         break;
 
       case RQ.CLIENT_MESSAGES.SYNC_APPLIED_RULES:
-        RQ.RuleExecutionHandler.syncCachedAppliedRules(
+        const newRules = RQ.RuleExecutionHandler.syncCachedAppliedRules(
```

Comment:
no need for this change. cachedRules for toplevel redirect are present in background itself.